### PR TITLE
refactor http client

### PR DESCRIPTION
**Pull request by tomerd**
motivation: prepare to consolidate downloader from TSC and http client from Basics

changes:
* add progress support to "execute"
* add authorization support to HTTPClientRequest::Options and HTTPClient::Configuration
* refactor call-sites to use new ways
* add and adopt tests

### CODE REVIEW

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by yim-lee** [comment]
Compilation error on Linux only? The PR looks good overall.

File: Sources/Basics/HTPClient+URLSession.swift [hunk 1]
```diff
@@ -8,28 +16,84 @@ import struct TSCUtility.Versioning
 import FoundationNetworking
 #endif
 
-public struct URLSessionHTTPClient: HTTPClientProtocol {
+public final class URLSessionHTTPClient: NSObject, HTTPClientProtocol {
```
Comment (tomerd) [inline Sources/Basics/HTPClient+URLSession.swift:19]: @neonichu  ptal 👀

File: Sources/Basics/HTTPClient.swift [hunk 2]
```diff
@@ -81,43 +95,68 @@ public struct HTTPClient: HTTPClientProtocol {
...
-                // handle retry strategy
-                if let retryDelay = self.shouldRetry(response: response, request: request, requestNumber: requestNumber) {
-                    self.diagnosticsEngine?.emit(warning: "\(request.url) failed, retrying in \(retryDelay)")
-                    // TODO: dedicated retry queue?
-                    return self.configuration.callbackQueue.asyncAfter(deadline: .now() + retryDelay) {
-                        self._execute(request: request, requestNumber: requestNumber + 1, callback: callback)
+        self.underlying(
+            request,
+            { received, expected in
+                if let max = request.options.maximumResponseSizeInBytes {
```
Comment (tomerd) [inline Sources/Basics/HTTPClient.swift:129]: @yim-lee this is to better support "maximumResponseSizeInBytes" at the response buffering  time

File: Tests/PackageCollectionsTests/JSONPackageCollectionProviderTests.swift [hunk 3]
```diff
@@ -275,15 +275,48 @@ class JSONPackageCollectionProviderTests: XCTestCase {
         })
     }
 
+    func testExceedsDownloadSizeLimitProgress() throws {
```
Comment (tomerd) [inline Tests/PackageCollectionsTests/JSONPackageCollectionProviderTests.swift:278]: @yim-lee  👀

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
> Compilation error on Linux only? The PR looks good overall.

yea seems like some of the nicer APIs are not supported on Linux, so had to do it the hard way... PR updated

**Review by neonichu — APPROVED** [review]

**Pull request merged**